controlplane/kubeadm/internal/cluster.go

```diff
@@ -214,23 +214,67 @@ func (c *cluster) generateEtcdTLSClientBundle() (*tls.Config, error) {
 	}, nil
 }
 
-// etcdIsHealthyResponse is a map of node provider IDs to the etcd health check that failed or nil if no checks failed.
-type etcdIsHealthyResponse map[string]error
+func (c *cluster) getControlPlaneNodes(ctx context.Context) (*corev1.NodeList, error) {
+	controlPlaneNodes := &corev1.NodeList{}
+	controlPlaneNodeLabels := map[string]string{
+		"node-role.kubernetes.io/master": "",
+	}
+
+	if err := c.client.List(ctx, controlPlaneNodes, client.MatchingLabels(controlPlaneNodeLabels)); err != nil {
+		return nil, err
+	}
+	return controlPlaneNodes, nil
+}
+
+// controlPlaneIsHealthy does a best effort check of the control plane components the kubeadm control plane cares about.
+// The return map is a map of node names as keys to error that that node encountered.
+// All nodes will exist in the map with nil errors if there were no errors for that node.
+func (c *cluster) controlPlaneIsHealthy(ctx context.Context) (map[string]error, error) {
+	controlPlaneNodes, err := c.getControlPlaneNodes(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	response := make(map[string]error)
+	for _, node := range controlPlaneNodes.Items {
+		name := node.Name
+		response[name] = nil
+		apiServerPodKey := types.NamespacedName{
+			Namespace: "kube-system",
+			Name:      staticPodName("kube-apiserver", name),
+		}
+		apiServerPod := &corev1.Pod{}
+		if err := c.client.Get(ctx, apiServerPodKey, apiServerPod); err != nil {
+			response[name] = err
+			continue
+		}
+		response[name] = checkStaticPodReadyCondition(apiServerPod)
+
+		controllerManagerPodKey := types.NamespacedName{
+			Namespace: "kube-system",
+			Name:      staticPodName("kube-controller-manager", name),
+		}
+		controllerManagerPod := &corev1.Pod{}
+		if err := c.client.Get(ctx, controllerManagerPodKey, controllerManagerPod); err != nil {
+			response[name] = err
+			continue
+		}
+		response[name] = checkStaticPodReadyCondition(controllerManagerPod)
```

Comment:
These are very similar, maybe put them in a function?

Reply:
i tried this but it ended up not really being much cleaner than this anyway. we can definitely reconsider if we add more pod checks though